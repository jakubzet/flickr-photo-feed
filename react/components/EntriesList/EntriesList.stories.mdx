import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import EntriesList from "./EntriesList";

<Meta
  title="Components|EntriesList"
  component={EntriesList}
  parameters={{ jest: ["EntriesList"] }}
/>

# EntriesList

EntriesList description placeholder.

<Preview>
  <Story name="default">
    <EntriesList />
  </Story>
</Preview>

<Props of={EntriesList} />