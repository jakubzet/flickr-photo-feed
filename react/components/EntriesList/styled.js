import styled from "styled-components";
import { rem, transparentize, darken, lighten } from "polished";
import { hideText } from "polished";

export const EntriesCategoryInfoWrapper = styled.div`
  color: ${p => p.theme.colors.white};
  background: ${p => p.theme.colors.persianBlue};
`;

export const EntriesCategoryInfo = styled.div`
  padding: 1rem;
  position: relative;

  ${p => p.theme.helpers.createWrapper()};

  @media screen and (min-width: ${p => p.theme.breakpoints.medium}) {
    flex-direction: row;
    align-items: center;
    padding: 2rem 1rem;
  }

  .EntriesCategoryInfo {
    display: flex;
    flex-direction: column;
    flex-basis: auto;
    opacity: ${p => (p.showContent ? 1 : 0)};
    visibility: ${p => (p.showContent ? "visible" : "hidden")};
    transition: opacity 400ms;

    &__primary {
      display: flex;
      flex-basis: auto;
      align-items: center;
      flex-grow: 1;
      font-family: ${p => p.theme.typography.type.secondary};
      font-weight: ${p => p.theme.typography.weight.bold};

      @media screen and (min-width: ${p => p.theme.breakpoints.medium}) {
        font-size: ${p => p.theme.typography.size[3]};
      }
    }

    &__secondary {
      flex-grow: 0;
      margin-top: 1rem;

      @media screen and (min-width: ${p => p.theme.breakpoints.medium}) {
        margin-top: 0;
      }
    }

    &__title {
      color: inherit;
    }

    &__link {
      background: url("../static/svg/link.svg") center center no-repeat;
      background-size: contain;
      height: ${rem(24)};
      width: ${rem(24)};
      margin-right: 0.5em;
      flex-shrink: 0;
      transition: transform ${p => p.theme.constants.transitionDuration};
      ${hideText()};

      &:hover {
        transform: scale(1.2);
      }
    }

    &__date {
      display: flex;
      flex-basis: auto;
      align-items: center;

      &:before {
        content: "";
        flex-shrink: 0;
        background: url("../static/svg/calendar.svg") center center no-repeat;
        background-size: cover;
        height: 1.5em;
        width: 1.5em;
        display: inline-block;
        margin-right: 0.5em;
      }
    }
  }
`;

export const EntriesList = styled.div`
  display: flex;
  flex-direction: column;
`;

export const Entries = styled.ul`
  flex-grow: 1;
`;

export const Entry = styled.li`
  padding: 1rem;
  border-top: dashed ${rem(1)}
    ${p => transparentize(0.75, p.theme.colors.shark)};

  &:nth-of-type(1) {
    border-top: none;
  }

  .Entry {
    &__container {
      display: flex;
      flex-direction: column;
      flex-basis: auto;
      overflow: hidden;
      ${p => p.theme.helpers.createWrapper()};

      @media screen and (min-width: ${p => p.theme.breakpoints.medium}) {
        flex-direction: row;
      }
    }

    &__wrapper {
      flex-grow: 0;
      overflow: hidden;
    }

    &__image-wrapper {
      width: 100%;
      height: auto;
      margin-bottom: 0.5rem;
      flex-shrink: 0;

      @media screen and (min-width: ${p => p.theme.breakpoints.medium}) {
        margin-right: 1rem;
        margin-bottom: 0;
        width: ${rem(340)};
        height: ${rem(256)};
      }

      & img {
        border-radius: ${rem(2)};
        object-fit: cover;
        width: inherit;
        height: inherit;
      }
    }

    &__title {
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      font-family: ${p => p.theme.typography.type.secondary};
      font-weight: ${p => p.theme.typography.weight.bold};
      font-size: ${p => p.theme.typography.size[2]};
      margin-bottom: 0.5em;
    }

    &__tags {
      display: flex;
      flex-basis: auto;
      flex-wrap: wrap;
      margin-bottom: 1em;

      li {
        & + li {
          margin-left: 0.5em;
        }
      }
    }

    &__details {
      display: flex;
      flex-basis: auto;
      flex-wrap: wrap;

      & > * {
        margin-right: 1em;
      }

      @media screen and (min-width: ${p => p.theme.breakpoints.large}) {
        flex-wrap: nowrap;
      }
    }

    &__date {
      width: 100%;
      order: 1;

      @media screen and (min-width: ${p => p.theme.breakpoints.large}) {
        width: auto;
        order: 3;
      }
    }

    &__author {
      order: 2;
    }

    &__link {
      order: 4;
    }
  }
`;

export const EntryTag = styled.div`
  background: ${p => p.theme.colors.neonCarrot};
  padding: 0.5em 1em;
  border-radius: ${rem(2)};
  text-transform: uppercase;
  font-size: 0.75em;
`;

export const EntriesFetchButton = styled.button`
  background: ${p => p.theme.colors.neonCarrot};
  color: ${p => p.theme.colors.shark};
  font-family: ${p => p.theme.typography.type.secondary};
  font-weight: ${p => p.theme.typography.weight.bold};
  transition: background ${p => p.theme.constants.transitionDuration};
  padding: 1rem;
  width: calc(100% - 1.5rem);
  text-transform: uppercase;
  border-radius: ${rem(2)};
  ${p => p.theme.helpers.createWrapper()};
  margin-bottom: 1rem;

  &:hover,
  &:active {
    cursor: pointer;
    background: ${p => darken(0.25, p.theme.colors.neonCarrot)};
  }

  &:disabled {
    cursor: not-allowed;
    background: ${p => lighten(0.25, p.theme.colors.neonCarrot)};
  }

  @media screen and (min-width: ${p => p.theme.breakpoints.medium}) {
    width: calc(100% - 2rem);
  }
`;
