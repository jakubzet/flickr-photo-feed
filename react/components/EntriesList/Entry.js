import React from "react";
import P from "prop-types";
import { prettifyDate } from "../../helpers";
import * as S from "./styled";

const defaultProps = {
  data: {}
};

const prepareTagsList = tagsStr => {
  return tagsStr
    .trim()
    .split(" ")
    .map((tag, idx) => {
      const tagKey = `tag_${idx}`;
      return (
        <li key={tagKey}>
          <S.EntryTag>{tag}</S.EntryTag>
        </li>
      );
    });
};

const prepareCreditsLink = (authorStr = 'sample@mail.com ("sampleNick")') => {
  const [mail, nickname] = authorStr.split(" ");

  const prettyNickname = nickname.match(/\(.+\)/i)[0].replace(/[(")]/gi, "");

  return <a href={`mailto:${mail}`}>{prettyNickname}</a>;
};

export const Entry = ({ data } = defaultProps) => {
  const hasTags = data.tags && data.tags.trim().length;

  return (
    <S.Entry>
      <article className="Entry__container">
        <div className="Entry__image-wrapper">
          <img src={data.media.m} alt={data.title} />
        </div>
        <div className="Entry__wrapper">
          <h2 className="Entry__title">{data.title}</h2>
          {hasTags && (
            <ul className="Entry__tags">{prepareTagsList(data.tags)}</ul>
          )}
          <div className="Entry__credits">
            Photo by: {prepareCreditsLink(data.author)}
          </div>
          <div className="Entry__details">
            <div className="Entry__date">
              Published: {prettifyDate(data.published)}
            </div>
            <div className="Entry__author">
              <a
                href={`https://www.flickr.com/photos/${data.author_id}`}
                target="_blank"
                rel="noopener noreferrer"
              >
                Author&apos;s page
              </a>
            </div>
            <div className="Entry__link">
              <a href={data.link} target="_blank" rel="noopener noreferrer">
                View on flickr
              </a>
            </div>
          </div>
        </div>
      </article>
    </S.Entry>
  );
};

Entry.propTypes = {
  data: P.shape({
    title: P.string.isRequired,
    link: P.string.isRequired,
    media: P.shape({
      m: P.string.isRequired
    }).isRequired,
    date_taken: P.string.isRequired,
    description: P.string.isRequired,
    published: P.string.isRequired,
    author: P.string.isRequired,
    author_id: P.string.isRequired,
    tags: P.string.isRequired
  })
};
