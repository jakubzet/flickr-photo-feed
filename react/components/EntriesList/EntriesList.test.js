import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { render } from "../../utils/reactTestingLibrary";
import { EntriesList } from "./EntriesList";

describe("EntriesList", () => {
  it("Renders", () => {
    const sampleEntries = [
      {
        title: "Lorem ipsum",
        link: "#",
        media: {
          m: "#"
        },
        date_taken: "Fri Dec 27 2019 01:14:14 GMT+0100",
        description: "Lorem ipsum",
        published: "Fri Dec 27 2019 01:14:14 GMT+0100",
        author: "Lorem ipsum",
        author_id: "id1",
        tags: "Lorem ipsum"
      }
    ];

    const { container, asFragment } = render(
      <Router>
        <EntriesList isFetching={false} entries={sampleEntries} />
      </Router>
    );

    expect(container).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });
});
