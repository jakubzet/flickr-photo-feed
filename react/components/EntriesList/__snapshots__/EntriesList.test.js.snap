// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntriesList Renders 1`] = `
<DocumentFragment>
  <div
    class="sc-htpNat eQVbpW"
  >
    <div
      class="sc-bdVaJa dabJse"
    >
      <div
        class="sc-bwzfXH bvKcmb"
      >
        <div
          class="EntriesCategoryInfo"
        >
          <div
            class="EntriesCategoryInfo__primary"
          >
            <a
              class="EntriesCategoryInfo__link"
            />
            <div
              class="EntriesCategoryInfo__title"
            />
          </div>
          <div
            class="EntriesCategoryInfo__secondary"
          >
            <div
              class="EntriesCategoryInfo__date"
            >
              Modified: 2rd January 1970 at 01:00
            </div>
          </div>
        </div>
      </div>
    </div>
    <ul
      class="sc-bxivhb khxFsg"
    >
      <li
        class="sc-EHOje kZZGZj"
      >
        <article
          class="Entry__container"
        >
          <a
            class="Entry__image-wrapper"
            href="/details/"
          >
            <img
              alt="Lorem ipsum"
              src="#"
            />
          </a>
          <div
            class="Entry__wrapper"
          >
            <a
              href="/details/"
            >
              <h2
                class="Entry__title"
              >
                Lorem ipsum
              </h2>
            </a>
            <ul
              class="Entry__tags"
            >
              <li>
                <a
                  class="sc-bZQynM hmQsdg"
                  href="https://www.flickr.com/search/?q=Lorem"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Lorem
                </a>
              </li>
              <li>
                <a
                  class="sc-bZQynM hmQsdg"
                  href="https://www.flickr.com/search/?q=ipsum"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  ipsum
                </a>
              </li>
            </ul>
            <div
              class="Entry__credits"
            >
              Photo by: 
              <a
                href="mailto:Lorem"
              >
                ipsum
              </a>
            </div>
            <div
              class="Entry__details"
            >
              <div
                class="Entry__date"
              >
                Published: 28th December 2019 at 01:14
              </div>
              <div
                class="Entry__author"
              >
                <a
                  href="https://www.flickr.com/photos/id1"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Author's page
                </a>
              </div>
              <div
                class="Entry__link"
              >
                <a
                  href="#"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  View on flickr
                </a>
              </div>
            </div>
          </div>
        </article>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
